### PR TITLE
Fix breakage on 0.5-dev.

### PR DESCRIPTION
**Pull request by yuyichao**
I'm not sure how much of these can go to `Compat.jl` so I just implement everything here....

c.c. @timholy 

Ref JuliaLang/julia#13803

### CODE REVIEW

**Comment by timholy** [comment]
I'm not sure those are material for Compat anyway since they were unexported functions. But if others feel differently, I'm fine with wherever they end up.

**Comment by yuyichao** [comment]
Oh SegFault, great .....

Also note that if we backport JuliaLang/julia#13803, the version check in this PR will needs to be tweaked....

**Comment by yuyichao** [comment]
Hmm. I couldn't reproduce the segfault but I'm getting some very strange errors in `reducedim.jl` test.

Depending on what I do before, I get either
```jl
ERROR: LoadError: UndefVarError: N not defined
 in _mapreducedim! at ./no file:4294967295
```
or
```jl
ERROR: LoadError: MethodError: `_nextract` has no method matching _nextract(::Type{Any}, ::Symbol, ::Expr)
Closest candidates are:
  _nextract(::Int64, ::Symbol, ::Expr)
  _nextract(::Int64, ::Symbol, ::Symbol)
 in func_for_method_checked at ./reflection.jl:248
```

when I try to call `@code_warntype Base._mapreducedim!(+, Base.AndFun(), Bool[], BitArray(10))`

This only happens if the module is precompiled but not if I add `--compilecache=no` or define the corresponding method at runtime again.

Can anyone reproduce this?

@vtjnash The method that is causing this issue directly is [here](https://github.com/JuliaStats/DataArrays.jl/blob/ef0ee793fa7c20eeb5ba7a2d0ebbc3130ea1f3ff/src/reducedim.jl#L10). Is there anything in it that looks suspicious for you in terms of precompilation compatibility? Also, is there a way I can check the `.ji` file and see if there's anything strange in there?

**Comment by yuyichao** [comment]
The first error `UndefVarError(:N)` is what I get when running the test locally without `--compilecache=no`.

**Comment by yuyichao** [comment]
Force pushed to trigger a travis rerun.

**Comment by yuyichao** [comment]
.... this time the error is the same one I see locally .....

https://travis-ci.org/JuliaStats/DataArrays.jl/jobs/88360029

**Comment by yuyichao** [comment]
The precompilation error is caused by https://github.com/JuliaLang/julia/pull/13677 ....

File: src/broadcast.jl [hunk 1]
```diff
@@ -1,7 +1,20 @@
 using DataArrays, Base.@get!
 using Base.Broadcast: bitcache_chunks, bitcache_size, dumpbitcache,
-                      promote_eltype, broadcast_shape, eltype_plus, type_minus, type_div,
-                      type_pow
+                      promote_eltype, broadcast_shape, eltype_plus
+
+if VERSION < v"0.5.0-dev+1016"
```
Comment (tkelman) [inline src/broadcast.jl]: can you use an `isdefined` check here rather than a version comparison?

Comment (yuyichao) [inline src/broadcast.jl]: Good point, will do when I'm back.

**Comment by yuyichao** [comment]
Updated the check to use `isdefined` so that it should work on 0.4 if we end up backporting JuliaLang/julia#13803 .

The nightly CI will fail until commit https://github.com/JuliaLang/julia/commit/5498570e2cb6a9b7065b7dc592c076b9b3125962 is included in the nightly binary (hopefully tomorrow...)

**Comment by tkelman** [comment]
should be up already: http://buildbot.e.ip.saba.us:8010/builders/package_tarball64/builds/145, hit restart if the `versioninfo()` looks too hold

**Comment by yuyichao** [comment]
You are right, the travis run fetch exactly the commit that has the fix. This should be ready to go then.

**Comment by tkelman** [comment]
@simonster et al, okay to merge this?

**Comment by tkelman** [comment]
Note that this needs to be tagged before https://github.com/JuliaLang/julia/pull/13803 can be considered for backporting. Not urgent as it's almost certainly going to be for 0.4.2, not 0.4.1.

**Pull request merged**